Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/RCFCaster.java

```diff
@@ -200,8 +212,66 @@ public ForecastDescriptor process(double[] inputPoint, long timestamp, int[] mis
     }
 
-    public void calibrate(Calibration calibration, double[] errorDeviations, RangeVector ranges) {
-        errorHandler.calibrate(calibration, errorDeviations, ranges);
+    public void calibrate(Calibration calibration, RangeVector ranges) {
+        errorHandler.calibrate(calibration, ranges);
+    }
+
+    @Override
+    public TimedRangeVector extrapolate(int horizon, boolean correct, double centrality) {
+        return this.extrapolate(calibrationMethod, horizon, correct, centrality);
+    }
+
+    public TimedRangeVector extrapolate(Calibration calibration, int horizon, boolean correct, double centrality) {
+        TimedRangeVector answer = super.extrapolate(horizon, correct, centrality);
+        calibrate(calibration, answer.rangeVector);
+        return answer;
     }
 
+    @Override
+    public List<AnomalyDescriptor> processSequentially(double[][] data, Function<AnomalyDescriptor, Boolean> filter) {
+        ArrayList<AnomalyDescriptor> answer = new ArrayList<>();
+        Function<AnomalyDescriptor, AnomalyDescriptor> function = (x) -> predictorCorrector.detect(x,
+                lastAnomalyDescriptor, forest);
+        if (data != null && data.length > 0) {
+            boolean cacheDisabled = (forest.getBoundingBoxCacheFraction() == 0);
+            try {
+                if (cacheDisabled) { // turn caching on temporarily
+                    forest.setBoundingBoxCacheFraction(1.0);
+                }
+                long timestamp = preprocessor.getInternalTimeStamp();
+                int length = preprocessor.getInputLength();
+                for (double[] point : data) {
+                    checkArgument(point.length == length, " nonuniform lengths ");
+                    ForecastDescriptor description = new ForecastDescriptor(point, timestamp++, forecastHorizon);
+                    description.setScoringStrategy(scoringStrategy);
+                    description = preprocessor.postProcess(predictorCorrector.detect(
+                            preprocessor.preProcess(description, lastAnomalyDescriptor, forest), lastAnomalyDescriptor,
+                            forest), lastAnomalyDescriptor, forest);
+                    if (saveDescriptor(description)) {
+                        lastAnomalyDescriptor = description.copyOf();
+                    }
+
+                    TimedRangeVector timedForecast = new TimedRangeVector(
+                            forest.getDimensions() * forecastHorizon / preprocessor.getShingleSize(), forecastHorizon);
+
+                    if (forest.isOutputReady()) {
+                        errorHandler.updateActuals(description.getCurrentInput(), description.getPostDeviations());
+                        errorHandler.augmentDescriptor(description);
+                        timedForecast = extrapolate(forecastHorizon);
+                        errorHandler.updateForecasts(timedForecast.rangeVector);
+                        description.setTimedForecast(timedForecast);
```

Comment:
Can you make this a method since another places also use these lines?

Reply:
Good suggestion. Actually we can take that a step further -- and now both the single step process and processSequentially call the exact same function and differ in the placement of the caching step. The main goal was to eliminate the repeated caching on/off and that is now clear. Done in https://github.com/aws/random-cut-forest-by-aws/pull/393/commits/e28e68a42ddc5a78d2a985f447e1b064f487488c